### PR TITLE
Adds initial persistent external storage support

### PR DESCRIPTION
**Pull request by lazypower**
To use on AWS EBS disk type:

juju deploy etcd
juju add-storage etcd/0 data=ebs,10g

To use on GCE PD Disk type:

juju deploy etcd
juju add-storage etcd/0 data=gce,10g

### CODE REVIEW

**Comment by marcoceppi** [comment]
README update would be nice. a few common callouts, like you have for AWS (ebs, ebs-ssd), GCE, then link to the Juju documentation.

Have not done a code review yet

File: reactive/etcd.py [hunk 1]
```diff
@@ -431,6 +432,87 @@ def passive_dismiss_context(cluster):
...
+    cmd = ['mkfs.ext4', device_info['location'], '-F']
+    hookenv.log('Creating filesystem on {}'.format(device_info['location']))
+    hookenv.log('With command: {}'.format(' '.join(cmd)))
+    check_call(cmd)
+
+    # Only attempt migration if directory exists
+    if os.path.isdir('/var/lib/etcd/default'):
+        hookenv.log('Detected existing data, migrating to new location.')
+        # Migrate any existing data
+        cmd = ['mkdir', '-p', '/mnt/etcd-migrate']
```
Comment (marcoceppi) [inline reactive/etcd.py]: Lets use the python module for this: https://docs.python.org/3/library/os.html?highlight=makedirs#os.makedirs

File: reactive/etcd.py [hunk 2]
```diff
@@ -431,6 +432,87 @@ def passive_dismiss_context(cluster):
...
+        hookenv.log('With command: {}'.format(' '.join(cmd)))
+        check_call(cmd)
+
+        mount_volume(block, '/mnt/etcd-migrate')
+
+        cmd = ['rsync', '-avz', '/var/lib/etcd/default', '/mnt/etcd-migrate/']
```
Comment (marcoceppi) [inline reactive/etcd.py]: `v` is excessive and `p` (partials) would be a good flag to add. If rsync is interrupted it'll pick up partials faster

File: reactive/etcd.py [hunk 3]
```diff
@@ -431,6 +432,87 @@ def passive_dismiss_context(cluster):
...
+        hookenv.log('With command: {}'.format(' '.join(cmd)))
+        check_call(cmd)
+
+        unmount_path('/mnt/etcd-migrate')
+
+    # halt etcd to perform the data-store migration
+    host.service_stop('etcd')
```
Comment (marcoceppi) [inline reactive/etcd.py:468]: Do we need to perform a second, quick sync after service_stop in the event data has changed since rsync and before halting?

**Review by marcoceppi — CHANGES_REQUESTED** [review]

**Comment by mbruzek** [comment]
I upgraded the etcd charm with a charm that I built from this branch. It upgraded cleanly and I was able to attach storage, as indicated in the top step. I do think we need the storage steps to the README.md file.

```
$ juju storage etcd
[Storage]  
Unit       Id      Location   Status    Message  
etcd/0     data/0  /dev/xvdf  attached 
```

```
# df -h
Filesystem      Size  Used Avail Use% Mounted on
udev            1.9G     0  1.9G   0% /dev
tmpfs           377M   11M  367M   3% /run
/dev/xvda1      7.8G  2.0G  5.4G  28% /
tmpfs           1.9G     0  1.9G   0% /dev/shm
tmpfs           5.0M     0  5.0M   0% /run/lock
tmpfs           1.9G     0  1.9G   0% /sys/fs/cgroup
/dev/xvdf       9.8G  341M  8.9G   4% /var/lib/etcd
tmpfs           377M     0  377M   0% /run/user/1000
```

LGTM +1

File: reactive/etcd.py (hunk 3 above)
Comment (lazypower) [reply reactive/etcd.py:468]: well, i could just move the service stop up. I was thinking the less time the service is offline the better, but i can see your point here, that we might be missing a few stray key-val sets.

**Pull request merged**